java-frontend/src/main/java/org/sonar/java/se/checks/InvariantReturnCheck.java

```diff
@@ -0,0 +1,174 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se.checks;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import org.sonar.check.Rule;
+import org.sonar.java.cfg.CFG;
+import org.sonar.java.collections.PMap;
+import org.sonar.java.se.CheckerContext;
+import org.sonar.java.se.constraint.Constraint;
+import org.sonar.java.se.constraint.ConstraintManager;
+import org.sonar.java.se.symbolicvalues.SymbolicValue;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Rule(key = "S3516")
+public class InvariantReturnCheck extends SECheck {
+
+  private static class MethodInvariantContext {
+    private final MethodTree methodTree;
+    private final Set<SymbolicValue> symbolicValues = new HashSet<>();
+    private final Multimap<Class<? extends Constraint>, Constraint> methodConstraints = ArrayListMultimap.create();
+    private final List<ReturnStatementTree> returnStatementTrees;
+    private final boolean methodToCheck;
+    private final boolean returnImmutableType;
+    private int endPaths = 0;
+    private boolean avoidRaisingConstraintIssue;
+
+    MethodInvariantContext(MethodTree methodTree) {
+      this.methodTree = methodTree;
+      TypeTree returnType = methodTree.returnType();
+      this.returnStatementTrees = extractReturnStatements(methodTree);
+      methodToCheck = !isConstructorOrVoid(methodTree, returnType) && returnStatementTrees.size() > 1;
+      returnImmutableType = methodToCheck && (returnType.symbolType().isPrimitive() || returnType.symbolType().is("java.lang.String"));
+    }
+
+    private static boolean isConstructorOrVoid(MethodTree methodTree, @Nullable TypeTree returnType) {
+      return methodTree.is(Tree.Kind.CONSTRUCTOR) || returnType.symbolType().isVoid();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "returnType" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)